Task30 Post /users initial commit

IlyaBaek opened the pull request.

IlyaBaek committed:
Task30 Post /users initial commit

EugenBorisik commented on Task10_Auth/src/main/java/ResponseWrapper.java:
```diff
@@ -1,8 +1,20 @@
 public class ResponseWrapper {
     private int responseCode;
-    private List<String> responseBody;
+    private List<String> responseBodyZipCodes;
+    private ArrayList<UsersDto> responseBodyUsers;
```
Let's store as single Object field

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/main/java/reseponsesDTO/UsersDto.java:
```diff
@@ -43,4 +47,42 @@ public String getZipCode() {
+        return Objects.hash(getName(), getUserSex());
+    }
+
+    public UsersDto createUserDtoObject(int age, String name, UsersDto.sex sex, String zipCode) {
+        UsersDto usersDto = new UsersDto();
+        usersDto.setAge(age);
+        usersDto.setName(name);
+        usersDto.setUserSex(sex);
+        usersDto.setZipCode(zipCode);
+        return usersDto;
+    }
+
+    public UsersDto createUserDtoObject(String name, UsersDto.sex sex) {
+        UsersDto usersDto = new UsersDto();
+        usersDto.setName(name);
+        usersDto.setUserSex(sex);
+        return usersDto;
+    }
```
Looks strange, make these methods as constructors or make them static

IlyaBaek replied:
How can I fill not all fields, If I will create a constructor with some default values.

For example, when I need only name and sex?

EugenBorisik replied:
You can have different ctors with different parameters (frankly speaking it is the main idea of ctors). Current solution: you're creating object to create another object of the same class, it is odd

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/main/java/reseponsesDTO/UsersDto.java:
```diff
@@ -2,12 +2,16 @@
     public sex userSex;
+    private String zipCode;
 
     public enum sex {FEMALE, MALE}
```
Enum must be named with upper letter

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/main/java/reseponsesDTO/UsersDto.java:
```diff
@@ -43,4 +47,42 @@ public String getZipCode() {
+        return usersDto;
+    }
+
+    public static UsersDto.sex randomSex()  {
```
Move this method in enum

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,108 @@
+        if (usersList.contains(userToAdd)) {
+            addedUser = usersList.get(usersList.indexOf(userToAdd));
+        }
+        UsersDto finalUserToAdd = userToAdd;
+        UsersDto finalAddedUser = addedUser;
```
These 2 variables are senseless

IlyaBaek replied:
Yes, but IDEA requires me to do that.

![image](https://user-images.githubusercontent.com/63204236/103171295-1f0bf300-485c-11eb-9b84-f73629a19341.png)

I have no idea how to solve it.

EugenBorisik replied:
Resolve issue with UsersDto class (see comments) and error will disappear.

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,108 @@
+        UsersDto finalUserToAdd = userToAdd;
+        UsersDto finalAddedUser = addedUser;
+        assertAll("Status code is 201 // user is added, zip code is removed",
+                () -> assertEquals(exactZipCodesCountBeforeCreation - 1, exactZipCodesCountAfterCreation),
```
Please verify that available zip codes list doesn't contain zip code of newly created user

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,108 @@
+
+        ResponseWrapper getZipCodesResponseAfterCreation = zipCodesClient.getZipCodes();
+        int exactZipCodesCountAfterCreation = Collections.frequency(getZipCodesResponseAfterCreation.getResponseBodyZipCodes(), exactZipCode);
+        UsersDto addedUser = new UsersDto();
+        ResponseWrapper getUsersResponse = usersClient.getUsers();
+        ArrayList<UsersDto> usersList = getUsersResponse.getResponseBodyUsers();
+        if (usersList.contains(userToAdd)) {
+            addedUser = usersList.get(usersList.indexOf(userToAdd));
+        }
```
I think that is extra step and you can use userToAdd as expected result. Your contains operation must be part of assertion.

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Test
+    public void createUserWithRequiredFieldsTest() {
```
See comments for previous test

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    @Test
+    public void createUserWithDuplicateNameAndSexTest() {
```
This test will fail if there are no users in app. Also, when you are trying to add user with unavailable zip code - you will get error.

IlyaBaek replied:
So, I have to check if there are some users created, and if there are no users I should fail this test with some message?

EugenBorisik replied:
You must be independent from existing users and just create them as precondition for your test. When you're testing that you can delete email in some email service - you're creating email before deletion

IlyaBaek replied:
done

EugenBorisik reviewed (changes requested):
Update code according to comments + implement last scenario from task 20 + cover additional scenario for create operation according to swagger.
Please pay your attention, that when your want to create object with random values - the best way is to encapsulate these random values in ctor and it doesn't make sense to pass them as explicit params

IlyaBaek committed:
Comments were fixed,
last scenario from task 20 was implemented.
Was added createUser scenario to cover 409 error.

EugenBorisik commented on Task10_Auth/src/main/java/ResponseWrapper.java:
```diff
@@ -1,22 +1,24 @@
 public class ResponseWrapper {
     private int responseCode;
-    private List<String> responseBody;
+    private ArrayList responseBody;
```
I still don't understand, why you're not using Object here

IlyaBaek replied:
resolved

EugenBorisik commented on Task10_Auth/src/main/java/UsersClient.java:
```diff
@@ -0,0 +1,79 @@
+        return responseWrapper;
+    }
+
+    public static void createUserIfNotExist(UsersClient usersClient) {
```
Trying to understand, why this method is static and others are non-static OR vice versa. Let's discuss in skype

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/main/java/ZipCodesClient.java:
```diff
@@ -58,4 +59,15 @@ public ResponseWrapper postZipCodes(List<String> body) {
         return responseWrapper;
     }
+
+    public static void createZipCodeIfNotExist(ZipCodesClient zipCodesClient){
```
The same question about static

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/main/java/reseponsesDTO/UsersDto.java:
```diff
@@ -43,4 +67,19 @@ public String getZipCode() {
+        if (this == o) return true;
+        if (!(o instanceof UsersDto)) return false;
+        UsersDto usersDto = (UsersDto) o;
+        if (getName() == null || getUserSex() == null) return false;
```
It makes sense to do the same for sex field too

IlyaBaek replied:
To do what?

EugenBorisik replied:
Will discuss in skype

IlyaBaek replied:
done,
added equals for all fields
and there is an additional check that required fields are not null

EugenBorisik commented on Task10_Auth/src/main/java/reseponsesDTO/UsersDto.java:
```diff
@@ -1,21 +1,45 @@
-    private int age;
-    private String name;
-    private String zipCode;
+    private int age = RandomUtils.nextInt(1, 99);
```
Lines 12, 13, 15: move initialization in default ctor (ctor without params)

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/test/java/ZipCodesTests.java:
```diff
@@ -57,12 +63,43 @@ public void addDuplicatesZipCodesAndCheckThereAreNoDuplicationsTest() {
+        List<UsersDto> usersList = new ArrayList<>(getUsersResponse.getResponseBody());
+        Random random = new Random();
+        List<String> newZipCodes = new ArrayList<>();
+        newZipCodes.add(usersList.get(random.nextInt(usersList.size())).getZipCode());
+        newZipCodes.add(usersList.get(random.nextInt(usersList.size())).getZipCode());
```
These zip codes can be equal to null and in this case test will be incorrect

IlyaBaek replied:
not relevant

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,106 @@
+    public void createUserWithoutRequiredField() {
+        String name = null;
+        UsersDto.Sex userSex = null;
+        UsersDto userToAdd = new UsersDto(name, userSex);
```
Can be simplified to 
`new UsersDto(null, null)`

IlyaBaek replied:
done

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Test
+    public void createUserWithIncorrectZipCodeTest() {
+        ResponseWrapper getZipCodesResponse = zipCodesClient.getZipCodes();
+        String zipCode;
+        do {
+            zipCode = RandomStringUtils.random(5, false, true);
+        } while (getZipCodesResponse.getResponseBody().contains(zipCode));
```
I think you can create ZipCodeGenerator class and paste these code in method in it, in this case you will encapsulate this code:
`ZipCodeGenerator.getRandomZipCode()`

IlyaBaek replied:
done

I've moved it to ZipCodesClient instead of a separate class.

EugenBorisik commented on Task10_Auth/src/test/java/CreateUsersTests.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Test
+    public void createUserWithRequiredFieldsTest() {
+        String name = RandomStringUtils.random(5, true, false);
+        UsersDto.Sex userSex = UsersDto.Sex.getRandomSex();
+        UsersDto userToAdd = new UsersDto(name, userSex);
```
If fields are random, then it makes sense to use ctor without params

IlyaBaek replied:
done

IlyaBaek committed:
Task30, PR changes
